Lib/test/test_asyncio/test_queues.py

```diff
@@ -522,5 +522,145 @@ class PriorityQueueJoinTests(_QueueJoinTestMixin, unittest.IsolatedAsyncioTestCa
 
 
+class _QueueShutdownTestMixin:
+    q_class = None
+
+    @staticmethod
+    async def _ping_awaitable(a):
+        try:
+            await asyncio.wait_for(asyncio.shield(a), 0.01)
+        except TimeoutError:
+            pass
+
+    def assertRaisesShutdown(self, msg="Didn't appear to shut-down queue"):
+        return self.assertRaises(asyncio.QueueShutDown, msg=msg)
+
+    async def test_format(self):
+        q = self.q_class()
+        q._is_shutdown = True
+        self.assertEqual(q._format(), 'maxsize=0 shutdown')
+
+    async def test_shutdown_empty(self):
+        q = self.q_class()
+        loop = asyncio.get_running_loop()
+        join_task = loop.create_task(q.join())
+        q.shutdown()
+
+        await self._ping_awaitable(join_task)
+        self.assertTrue(join_task.done())
+        with self.assertRaisesShutdown():
+            await q.put("data")
+        with self.assertRaisesShutdown():
+            await q.get()
+        await join_task
+
+    async def test_shutdown_nonempty(self):
+        q = self.q_class()
+        loop = asyncio.get_running_loop()
+        q.put_nowait("data")
+        join_task = loop.create_task(q.join())
+        q.shutdown()
+
+        self.assertEqual(await q.get(), "data")
+
+        await self._ping_awaitable(join_task)
+        self.assertFalse(join_task.done())
+
+        with self.assertRaisesShutdown():
+            await q.put("data")
+        with self.assertRaisesShutdown():
+            q.put_nowait("data")
+
+        with self.assertRaisesShutdown():
+            await q.get()
+        with self.assertRaisesShutdown():
+            q.get_nowait()
+
+        q.task_done()
+
+        await self._ping_awaitable(join_task)
+        self.assertTrue(join_task.done())
+        await join_task
+
+    async def test_shutdown_immediate(self):
+        q = self.q_class()
+        loop = asyncio.get_running_loop()
+        q.put_nowait("data")
+        join_task = loop.create_task(q.join())
+        q.shutdown(immediate=True)
+
+        await self._ping_awaitable(join_task)
+        self.assertTrue(join_task.done())
+
+        with self.assertRaisesShutdown():
+            await q.put("data")
+        with self.assertRaisesShutdown():
+            q.put_nowait("data")
+
+        with self.assertRaisesShutdown():
+            await q.get()
+        with self.assertRaisesShutdown():
+            q.get_nowait()
+
+        with self.assertRaises(
+            ValueError, msg="Didn't appear to mark all tasks done"
+        ):
+            q.task_done()
+
+        await self._ping_awaitable(join_task)
+        self.assertTrue(join_task.done())
+        await join_task
+
+    async def test_shutdown_immediate_with_unfinished(self):
+        q = self.q_class()
+        loop = asyncio.get_running_loop()
+        q.put_nowait("data")
+        q.put_nowait("data")
+        join_task = loop.create_task(q.join())
+        self.assertEqual(await q.get(), "data")
+        q.shutdown(immediate=True)
+
```

Comment:
Could you please,add a comment here about the queue state just after the shutdown ?
No item in the queue, self._unfinished_tasks is stiil set to 1.

Reply:
I've added a comment indicating (expected) changes to unfinished-tasks, but for number of items I've instead added a check